### ISSUE TITLE
Make BIR better capture structure of control flow

### ISSUE THREAD

**Issue by jclark** [ballerina-platform/nballerina#921]
Currently BIR represents control flow as just a control flow graph. This is all that is needed for conversion to LLVM, but loses information about the structure that was present in the source.  This makes several things more difficult than they could be:

- verification of conditional type narrowing (as proposed here #883, or as currently in the conditional type narrowing instruction)
- construction of SSA - if a backend wants to build an SSA representation of the program for optimization purposes, it would be easier to do this if control flow were represented in a structured way (see [Single-pass generation of static single-assignment form for structured languages])(https://dl.acm.org/doi/abs/10.1145/197320.197331)
- generation of WASM - this now has to reconstruct the structure of the control flow from the CFG, where this is already explicit in the syntax tree
- it allows irreducible flow graphs - backends ought to support any legal BIR, but BIR allows irreducible flow graphs while Ballerina does not

We want to do this as well as #883.

**Comment by jclark** [comment]
GraalVM IR has an interesting approach to representing control flow https://ssw.jku.at/General/Staff/GD/APPLC-2013-paper_12.pdf

**Comment by jclark** [comment]
I think the general approach that can work is to allow a block to have properties, and to have constraints about how blocks with particular properties are connected in the graph, which can be enforced in the verifier.

Consumers of the BIR who just want a CFG can ignore the properties.  Those that need structural information can may attention to the properties.

The design goal for the properties and constraints would be to address the points mentioned in the first comment
- facilitate verification using #883 
- facilitate SSA construction
- facilitate WASM generation
- prohibit irreducible graphs

**Comment by jclark** [comment]
Ideas:

- edges from conditional branches are all forward
- forward branches form a DAG
- property on a block saying it's the loop head
- property on a block says terminating branch is backward
- backward branch only allowed to loop head
- property on a block saying it's the exit block of a loop (where break would go to)
- one exit block per loop
- a block is in a loop if it is reachable from the loop head without going through the loop's exit block

(This is too strong for `on fail`)

**Comment by jclark** [comment]
Probably better to have some of these properties on the branch instruction, specifically whether the branch is backwards.

**Comment by jclark** [comment]
I am not clear about whether we need more structure for if/else.

**Comment by jclark** [comment]
This paper [Johnson and Pingali. Dependence-based program analysis](http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.38.7952&rep=rep1&type=pdf) has the concept of single-entry/single-exit regions.  Source statements have corresponding single-entry/single-exit regions, and nesting of source statements turns into nesting of these regions.

There are a number of complications we have to deal with:
- return
- panic/trap
- fail/on fail

Putting those on one side, the idea would be that BIR would
- identify single-entry/single-exit regions
- say when a single-entry/single-exit region is a loop
- say when a branch is backwards - a backwards branch would be allowed only to a single-entry/single-exit region that is a loop

For return and panic/trap, we have specific instructions in the BIR. `return` will be a little more complicated with named workers, in that a return in a given function can jump to different places depending on which worker it is in.

We can conceptualize this by thinking of the exit of a single-entry/single-exit region as having up to three lanes: one for the normal case, one for fail and one for panic.

We can handle `panic` by storing an association between the block for the exit for the region and the block to which panics within the region jump.

Similarly for fail, the BIR should
- mark the branch as being for a fail
- record an association between the regular exit for the block inside the on-fail and the block that handles the fail 

The rule would then be that a fail branch with B has to jump to a block associated with the exit block of a region containing B.

**Comment by jclark** [comment]
Can we go as far as saying that any merge block (two or more predecessors) is the entry or exit block of a single-entry/single-exit region?  Or maybe just that TypeMerge insns (#883) are in the exit block?  We would also need TypeMerge insns in the block  `on fail` (tied with an exit block).

**Comment by manuranga** [comment]
In this scheme, will we be able to easily recover the structured flow of logical expressions?. eg:

```ballerina
int|float|string a = …;
if a is int || a is float {
   f(a); 
}
```

![short-circuit-or](https://user-images.githubusercontent.com/1686124/153807519-52d8bafc-c1da-44ab-8292-1d6fa9d61588.svg)

**Comment by jclark** [comment]
We could model the condition as a region with one entry and two exits (one for true and one for false); we would then have the constraint that in such a region, assignments (other than to temp registers) is disallowed.

The two exits allow us to model the separate flow of type information implied for true and false cases  an if/else statement.

But I think the only case where this matters is WASM generation.

**Comment by poorna2152** [comment]
> In this scheme, will we be able to easily recover the structured flow of logical expressions?

For the above representation following is the output generated [using the relooper](https://gist.github.com/poorna2152/985ae28df808895045c82930c46cd047) ,

- WAT 

```wasm
(module
  (import "console" "log" (func $f (param i64)))
  (import "Number" "isInteger" (func $isInt (param i64) (result i32)))
  (import "Number" "isFloat" (func $isFloat (param i64) (result i32)))
  (export "printOr" (func $printOr))
  (func $printOr (param $0 i64)
    (block $block$4$break
      (block $block$2$break
        (block)
          (if
            (call $isInt
              (local.get $0))
            (block
              (br $block$2$break))
            (block
              (block)
              (if
                (call $isFloat
                  (local.get $0))
                (block
                  (br $block$2$break))
                (block
                  (br $block$4$break))))))
      (block
        (call $f
          (local.get $0))
        (block
          (br $block$4$break))))
    (block
      (return))))


```
- Relooper Block Structure 
![Relooper Block structure](https://user-images.githubusercontent.com/58025190/153824128-51b71162-733c-4c91-8136-8579b15ce35e.png)

**Comment by jclark** [comment]
@poorna2152 When you format WAT, please

* don't put the close parenthesis on a separate line and
* indent by more than 1 space.

**Comment by poorna2152** [comment]
> @poorna2152 When you format WAT, please
> 
> * don't put the close parenthesis on a separate line and
> * indent by more than 1 space.

Updated the comment.

**Comment by manuranga** [comment]
Optimized  with binaryen
```wat
(module
    (type $i64_=>_none (func (param i64)))
    (type $i64_=>_i32 (func (param i64) (result i32)))
    (import "console" "log" (func $f (param i64)))
    (import "Number" "isInteger" (func $isInteger (param i64) (result i32)))
    (import "Number" "isFloat" (func $isFloat (param i64) (result i32)))
    (export "printOr" (func $0))
    (func $0 (param $0 i64)
        (block $label$1
            (if
                (i32.eqz
                    (call $isInteger
                        (local.get $0)))
                (br_if $label$1
                    (i32.eqz
                        (call $isFloat
                            (local.get $0)))))
            (call $f
                (local.get $0)))))
```

**Comment by manuranga** [comment]
![short-circuit-or-opt](https://user-images.githubusercontent.com/1686124/153906162-20db2d66-2800-47e9-8460-40cf9394d456.svg)

**Comment by jclark** [comment]
Note that these regions may correspond to scopes for variables needed for debug info.

**Comment by jclark** [comment]
Concretely, we add a field to bir:FunctionCode

```
  Region[] regions;
```

where Region is

```
type Region readonly & record {|
    Label entry;
    Label? exit;
    Label? parent;
    boolean isLoop;
|};
```

Then we generate these regions in stmt.bal and verify them in verify.bal.

Also add field to unconditional branch saying whether its backwards.

**Comment by heshanpadmasiri** [comment]
Resolved with #947
